### ISSUE TITLE
Hello, args.train_data is a folder that contains a lot of tar files. Could I use the following function to load these data successfully? There are many image pairs in each tar file, and the image and text of each image pair have the same name. Hope to get your reply, thanks!!

### ISSUE THREAD

**Issue by leo23ui** [webdataset/webdataset#426]
Hello, args.train_data is a folder that contains a lot of tar files. Could I use the following function to load these data successfully? There are many image pairs in each tar file, and the image and text of each image pair have the same name. Hope to get your reply, thanks!!

def get_wds_dataset(args, preprocess_img, is_train, epoch=0, floor=False, tokenizer=None):

    def get_tar_files_from_folder(folder):
        return sorted(glob.glob(os.path.join(folder, "*.tar")))

    if is_train:
        if os.path.isdir(args.train_data): 
            input_shards = get_tar_files_from_folder(args.train_data)
        else:
            input_shards = args.train_data  
    else:
        if os.path.isdir(args.val_data):  
            input_shards = get_tar_files_from_folder(args.val_data)
        else:
            input_shards = args.val_data
    print(f"input_shards type: {type(input_shards)}")

    assert input_shards is not None
    resampled = getattr(args, 'dataset_resampled', False) and is_train

    num_shards = None
    if is_train:
        if args.train_num_samples is not None:
            num_samples = args.train_num_samples
        else:
            num_samples, num_shards = get_dataset_size(input_shards)
            if not num_samples:
                raise RuntimeError(
                    'Currently, the number of dataset samples must be specified for the training dataset. '
                    'Please specify it via `--train-num-samples` if no dataset length info is present.')
    else:
        num_samples = args.val_num_samples or 0

    shared_epoch = SharedEpoch(epoch=epoch)
        pipeline = [wds.SimpleShardList(input_shards)]

    if is_train:
        if not resampled:
            pipeline.extend([
                detshuffle2(
                    bufsize=_SHARD_SHUFFLE_SIZE,
                    initial=_SHARD_SHUFFLE_INITIAL,
                    seed=args.seed,
                    epoch=shared_epoch,
                ),
                wds.split_by_node,
                wds.split_by_worker,
            ])
        pipeline.extend([
            tarfile_to_samples_nothrow,
            wds.shuffle(
                bufsize=_SAMPLE_SHUFFLE_SIZE,
                initial=_SAMPLE_SHUFFLE_INITIAL,
            ),
        ])
    else:
        pipeline.extend([
            wds.split_by_worker,
            wds.tarfile_to_samples(handler=log_and_continue),
        ])
    pipeline.extend([
        wds.select(filter_no_caption_or_no_image),
        wds.decode("pilrgb", handler=log_and_continue),
        wds.rename(image="jpg;png;jpeg;webp", text="txt"),
        wds.map(check_filenames),

        wds.map_dict(image=preprocess_img,
                     text=lambda text: tokenizer(text)[0]),
        wds.to_tuple("image", "text"),
        wds.batched(args.batch_size, partial=not is_train)
    ])

    dataset = wds.DataPipeline(*pipeline)

    if is_train:
        if not resampled:
            num_shards = num_shards or len(input_shards)
            assert num_shards >= args.workers * args.world_size, 'number of shards must be >= total workers'
        round_fn = math.floor if floor else math.ceil
        global_batch_size = args.batch_size * args.world_size
        num_batches = round_fn(num_samples / global_batch_size)
        num_workers = max(1, args.workers)
        num_worker_batches = round_fn(num_batches / num_workers)
        num_batches = num_worker_batches * num_workers
        num_samples = num_batches * global_batch_size
        dataset = dataset.with_epoch(num_worker_batches)
    else:
        num_batches = math.ceil(num_samples / args.batch_size)

    dataloader = wds.WebLoader(
        dataset,
        batch_size=None,
        shuffle=False,
        num_workers=args.workers,
        persistent_workers=args.workers > 0,
    )

    dataloader.num_batches = num_batches
    dataloader.num_samples = num_samples

**Comment by tmbdev** [comment]
It sounds like the question you have is really about how to deal with multiple tar files that contain duplicate file names.

The short answer is: you're best off repackaging the data into tar files with unique names. That makes debugging and error analysis much easier.

If you really must use data like this in its original form, you might want to consider renaming the files before the `tarfiles_to_samples` function inside the pipeline; you could simply incorporate part of the tar file name into the file name.
